subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/InputToPropertyMapping.java

```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.history.changes;
+
+import com.google.common.collect.ImmutableMap;
+
+public interface InputToPropertyMapping {
+    ImmutableMap<Object, String> getInputToPropertyNames();
+
+}
```

Comment:
Why this separate interface? Nobody else implements it. I think this method should instead be a visitor like `visitOutputProperties` inside `UnitOfWork`.

Reply:
I created the separate interface to remove the package cycle between `execution` and `execution.history.changes`. The `execution.history.changes` is referenced by `execution`, but not the other way round.
Using a visit method sounds good, though it doesn't help with solving the package cycle.

Maybe it would be better for the `ExecuteStep` to create the `InputChanges` object, when required, and pass it to the execute method. For this to work, the unit of work needs to expose whether it uses incremental inputs or not. This information is easily available for artifact transforms, for tasks we may need to query the `TaskClassInfoStore`, the task actions or pass the information somehow to the task executer. At some point, I added a field on `TaskInternal`, though I would really prefer not to.

Reply:
I think what you suggest is on the right track. I'm not sure if it should be part of `ExecuteStep` or it should be a separate step, but we can always extract it later if it looks better.

Reply:
I added it to `ExecuteStep` now.